torch/csrc/jit/tensorexpr/DesignOverview.md

```diff
@@ -0,0 +1,96 @@
+
+    Tensor* x = Compute(
+        "x",
+        {{M, "m1"}, {N, "n1"}, {K, "k1"}},
+        [&](const Var& m, const Var& n, const Var& k) {
+          return a_buf(m, n) * b_buf(n, k);
+        });
+    Tensor* y = ...;
+    Tensor* z = ...;
+    std::vector<Tensor*> tensors_to_compute = {x, z}; // Tensor y might be used in x or z - in this case it will also be computed. 
+```
+
+## Step 2: Create schedule for the tensor expressions:
+```
+   Schedule s(tensors_to_compute);
+```
+This constructs a tree-like data structure (`TensorExprNode`) representing loop nests for the given tensor computation.
+A node in this IR is either a loop-axis(LoopAxis) or a tensor expression (`TensorExprOp`).
+If it is a loop-axis, it also contains children that again might be either a loop-axes or a tensor expression, and so on.
+If it is a tensor-expression, it is lowered to a statement (`Stmt`). Currently, it just means that we're creating a `Store` for every tensor-expression. We also keep a pointer to the original tensor expression.
+It could look like this:
+```
+loop-axis i
+  loop-axis j
+    Store(to: a[i, j], what: x[i] + y[j])
+loop-axis k
+  loop-axis l
+    Store(to: b[k, l], what: a[i, j] + 1)
+    loop-axis m
+      Store(to: c[k,l,m], what: b[k,l] + z[m])
+```
+
+## Step 3: Apply scheduling primitives
+Scheduling primitives mutate the tree structure: they can create or remove loop-axis, replace statements with other statements (updates `element_stmt` for each affected tensor expression) or remove them. The transformations also record the history.
+The output of this step is a modified tree-like structure (same format as in step 2).
+
+## Step 4: Lower the tree structure to statements.
+This step creates a `For` statement for each loop-axis and emits `element_stmt` for bodies of the loops.
+
+## Step 5: Pass the final statement for codegen (LLVM/CUDA/IREval)
+Codegen is implemented as an IR visitor over the statements produced in the previous step.
+
+# Tensor Expressions Language
+There are several core concepts in the Tensor Expression engine, this section defines them and shows how they connect to each other.
+
+## Expr
+Expr represents a node in the abstract syntax tree of a tensor expression. Leaf nodes in such tree are either a symbolic variable (`Var`), a constant (`IntImm` or `FloatImm`), `Buffer`, or a `Tensor`. Non-leaf nodes refer to other expressions and represent various operations. E.g. `Add` has two operands: `lhs` and `rhs`, both of which are also `Expr`.
+
+## Tensor
+`Tensor` is a bundle of
+1) a variable `Var` defining which tensor this `Tensor` expression is describing
+2) a list of indices `args` (each of them is `Var`)
+3) a list of expressions for dimensions `dims` (each of them is `Expr`)
+4) a computational expression `body` (of `Expr` type)
+
+## Buffer
+`Buffer`s are essentially `Tensor`s without a `body` - they represent an indexed access to something not defined in the tensor-expression.
```

Comment:
Minor: They represent a symbolic access to "tensors" that is outside the tensor-expression system.

Reply:
Right, that's what I tried to say, I'll borrow your version :)